### PR TITLE
registration, login

### PR DESCRIPTION
**Pull request by vladimirov-s**

### CODE REVIEW

File: src/controllers/user-controller.js [hunk 1]
```diff
@@ -0,0 +1,78 @@
+const User = require('./../models/user-schema');
+const bcrypt = require('bcrypt');
+const jwt = require('jsonwebtoken');
+const { validationResult } = require('express-validator');
+const { secret } = require('./../../config')
```
Comment (baklanovairina) [inline src/controllers/user-controller.js:5]: ;

File: src/controllers/user-controller.js [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+const HASH_ROUDS = 10;
+
+const generateAccessToken = (id) => {
+  const payload = {
+    id
+  }
+  return jwt.sign(payload, secret, { expiresIn: "24h" });
+}
+
+
```
Comment (baklanovairina) [inline src/controllers/user-controller.js:15]: remove extra empty line

File: src/controllers/user-controller.js [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+class UserController {
+  async registration(req, res) {
+    try {
+      const errors = validationResult(req);
+
+      if (!errors.isEmpty()) {
+        return res.status(400).json({ data: 'Ащибка при регистрации', errors });
```
Comment (baklanovairina) [inline src/controllers/user-controller.js:22]: please write everywhere without mistakes

File: src/controllers/user-controller.js [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+      }
+
+      const { email, password } = req.body;
```
Comment (baklanovairina) [inline src/controllers/user-controller.js:25]: and do we use email, not login?

File: src/controllers/user-controller.js [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+      const candidate = await User.findOne({ email: email });
+
+      if (candidate) {
+        return res.status(400).json({ message: "Пользователь с таким именем уже существует" })
```
Comment (baklanovairina) [inline src/controllers/user-controller.js:29]: ;

File: src/controllers/user-controller.js [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+      }
+
+      const hashPassword = bcrypt.hashSync(password, HASH_ROUDS);
+      const user = new User({ email: email, password: hashPassword });
+      await user.save();
+
+      return res.json({ data: 'Пользователь успешно зарегестрирован.' })
```
Comment (baklanovairina) [inline src/controllers/user-controller.js:36]: misspelled "зарегистрирован"

File: src/controllers/user-controller.js [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    } catch (err) {
+      console.log(err);
+      res.status(400).json({ data: "Registration error" })
+    }
+  }
+
+
```
Comment (baklanovairina) [inline src/controllers/user-controller.js:43]: extra empty line

File: src/controllers/user-controller.js [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+  async login(req, res) {
+    try {
+      const { email, password } = req.body;
```
Comment (baklanovairina) [inline src/controllers/user-controller.js:46]: why email?

File: src/controllers/user-controller.js [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+      const user = await User.findOne({ email });
+
+      if (!user) {
+        return res.status(400).json({ data: `Пользователь ${email} не найден` });
+      }
+
+      const validPassword = bcrypt.compareSync(password, user.password);
+      if (!validPassword) {
+        return res.status(400).json({ data: `Введен не верный пароль` })
```
Comment (baklanovairina) [inline src/controllers/user-controller.js:55]: ;

File: src/controllers/user-controller.js [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+      const token = generateAccessToken(user._id);
+      return res.json({ token });
+
+    } catch (err) {
+      console.log(e);
+      res.status(400).json({ data: "Login error" })
+    }
+  }
+
+
```
Comment (baklanovairina) [inline src/controllers/user-controller.js:66]: extra empty line

File: src/controllers/user-controller.js [hunk 11]
```diff
@@ -0,0 +1,78 @@
+const User = require('./../models/user-schema');
+const bcrypt = require('bcrypt');
+const jwt = require('jsonwebtoken');
+const { validationResult } = require('express-validator');
+const { secret } = require('./../../config')
+const HASH_ROUDS = 10;
+
+const generateAccessToken = (id) => {
```
Comment (baklanovairina) [inline src/controllers/user-controller.js:8]: it is better to move work with tokens into a separate file and class

File: src/controllers/user-controller.js (hunk 6 above)
Comment (baklanovairina) [inline src/controllers/user-controller.js:36]: why don't you generate a token here and log in immediately after registration?

File: src/controllers/user-controller.js [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+  const payload = {
+    id
+  }
+  return jwt.sign(payload, secret, { expiresIn: "24h" });
```
Comment (baklanovairina) [inline src/controllers/user-controller.js:12]: why don't you use refreshToken to refresh accessToken?

File: src/controllers/user-controller.js [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+      return res.json({ token });
+
+    } catch (err) {
+      console.log(e);
+      res.status(400).json({ data: "Login error" })
+    }
+  }
+
+
+  async getUsers(req, res, next) {
```
Comment (baklanovairina) [inline src/controllers/user-controller.js:67]: why getUsers?

File: src/models/user-schema.js [hunk 14]
```diff
@@ -0,0 +1,7 @@
+const { Schema, model } = require('mongoose');
```
Comment (baklanovairina) [inline src/models/user-schema.js:1]: what is a model?

File: src/models/user-schema.js [hunk 15]
```diff
@@ -0,0 +1,7 @@
...
+const User = new Schema({
+  email: { type: String, unique: true, required: true },
```
Comment (baklanovairina) [inline src/models/user-schema.js:3]:
```
  login: { 
    type: String, 
    unique: true, 
    required: true 
  },
```

File: src/models/user-schema.js [hunk 16]
```diff
@@ -0,0 +1,7 @@
...
+  password: { type: String, required: true }
```
Comment (baklanovairina) [inline src/models/user-schema.js:4]: likewise

File: src/service/user-service.js [hunk 17]
```diff
@@ -0,0 +1,21 @@
+const UserModel = require('./../models/user-schema');
```
Comment (baklanovairina) [inline src/service/user-service.js:1]: use the name given in "user-schema"

File: src/models/user-schema.js (hunk 14 above)
Comment (baklanovairina) [inline src/models/user-schema.js:1]: the file name "user-schema.js" can be abbreviated to "user.js"

File: src/service/user-service.js [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+const bcrypt = require('bcrypt');
+const uuid = require('uuid');
+const tokenService = require('./token-service');
+
+class UserService {
+  async registration(email, password) {
+    const candidate = await UserModel.findOne({ email })
```
Comment (baklanovairina) [inline src/service/user-service.js:8]: why are you writing the same thing here as in controllers?

File: src/controllers/user-controller.js (hunk 4 above)
Comment (vladimirov-s) [reply src/controllers/user-controller.js:25]: which is a differences?? Tommorrow it will be an email or alias

File: src/models/user-schema.js (hunk 14 above)
Comment (vladimirov-s) [reply src/models/user-schema.js:1]: member BEM

File: src/controllers/user-controller.js (hunk 4 above)
Comment (baklanovairina) [reply src/controllers/user-controller.js:33]: Because you work with a login, not an email. This can confuse a person, since the work of email and login may differ. Plus, if the program expands and a new input is created, for example, for an email, then this email of yours will have to be rewritten.

File: src/controllers/user-controller.js [hunk 19]
```diff
@@ -0,0 +1,113 @@
+const User = require("./../models/user-schema");
+const bcrypt = require("bcrypt");
+const jwt = require("jsonwebtoken");
+const {
+  validationResult,
```
Comment (baklanovairina) [inline src/controllers/user-controller.js:5]: What for? You have only one parameter. Such an entry is not needed for a single value.

File: src/controllers/user-controller.js [hunk 20]
```diff
@@ -0,0 +1,113 @@
...
+} = require("express-validator");
+const {
```
Comment (baklanovairina) [inline src/controllers/user-controller.js:7]: What for? You have only one parameter. Such an entry is not needed for a single value.

File: src/controllers/user-controller.js [hunk 21]
```diff
@@ -0,0 +1,113 @@
...
+        HASH_ROUDS
+      );
+      const user = new User({
+        email: email,
+        password: hashPassword,
+      });
+      await user.save();
+
+      return res.json({
+        data: "Пользователь успешно зарегестрирован.",
```
Comment (baklanovairina) [inline src/controllers/user-controller.js:58]: Please correct any typos

File: src/controllers/user-controller.js (hunk 13 above)
Comment (baklanovairina) [reply src/controllers/user-controller.js:103]: And again, why do you need to get all users??

File: src/models/user-schema.js [hunk 22]
```diff
@@ -0,0 +1,17 @@
+const {
+  Schema,
+  model,
```
Comment (baklanovairina) [inline src/models/user-schema.js:3]: Why are you unloading the model? mongoose does not require this

File: src/routes/router.js [hunk 23]
```diff
@@ -0,0 +1,14 @@
+const Router = require('express').Router;
+const userController = require('../controllers/user-controller');
+const router = new Router();
+const { check } = require('express-validator');
+const authMiddleWare = require('./../../middleWaree/authMiddleWare');
+
+router.post('/registration', [
+  check('email', 'Имя пользователя не может быть пустым').notEmpty(),
```
Comment (baklanovairina) [inline src/routes/router.js:8]: The login also has validation: the string length is at least 6 characters

File: src/routes/router.js [hunk 24]
```diff
@@ -0,0 +1,14 @@
...
+  check('password', 'Пароль должен быть не меньше 4 и не больше 10 символов').isLength({ min: 4, max: 10 })
```
Comment (baklanovairina) [inline src/routes/router.js:9]: password must be at least 6 characters long

File: src/controllers/user-controller.js (hunk 20 above)
Comment (vladimirov-s) [reply src/controllers/user-controller.js:7]: When u check the work u send incorrect data to the backend. If there is not a validate the script brokes. That is why u should have definition such about functionality and checking methods

File: src/models/user-schema.js (hunk 22 above)
Comment (vladimirov-s) [reply src/models/user-schema.js:3]: For manage the database.
Such it works

File: src/routes/router.js (hunk 24 above)
Comment (vladimirov-s) [reply src/routes/router.js:9]: which is a difference?????

Comment (baklanovairina) [reply src/routes/router.js:9]: and such that you have a technical task that must be carried out exactly in accordance 
Or is it just designed that way? And you will also argue with the customer and say what's the difference?

File: src/models/user-schema.js (hunk 22 above)
Comment (baklanovairina) [reply src/models/user-schema.js:3]: It works without it too. Why write this?

File: src/controllers/user-controller.js (hunk 20 above)
Comment (baklanovairina) [reply src/controllers/user-controller.js:7]: I didn't ask about that. I asked why write in 3 lines if you need to write in one

File: src/controllers/user-controller.js (hunk 19 above)
Comment (vladimirov-s) [reply src/controllers/user-controller.js:5]: So, it works code formatter.

File: src/controllers/user-controller.js (hunk 20 above)
Comment (vladimirov-s) [reply src/controllers/user-controller.js:7]: If I set another settings- you bring notes on another places

File: src/controllers/user-controller.js (hunk 13 above)
Comment (baklanovairina) [reply src/controllers/user-controller.js:103]: Can you at least answer why you need this function?

File: src/models/user-schema.js [hunk 25]
```diff
@@ -0,0 +1,13 @@
+const User = new Schema({
```
Comment (baklanovairina) [inline src/models/user-schema.js:1]: why did you delete everything? It won't work for you without schema import.

Comment (vladimirov-s) [reply src/models/user-schema.js:1]: > Can you at least answer why you need this function?

So I have answered eyt - it is for manage

File: src/controllers/user.controller.js [hunk 26]
```diff
@@ -0,0 +1,87 @@
+const bcrypt = require("bcrypt");
+require("dotenv").config();
```
[the remark this answers is not in the document]
Comment (vladimirov-s) [reply src/controllers/user.controller.js:2]: we deprecate the config. So secret I should place in .env

File: index.js [hunk 27]
```diff
@@ -0,0 +1,27 @@
...
+
+const start = async () => {
+  try {
+    await mongoose.connect(process.env.URL, {
+      useNewUrlParser: true,
+      useUnifiedTopology: true,
+    });
+    app.listen(port, () => console.log(`listen port-${port}`));
+  } catch (err) {
+    console.log(err);
```
Comment (baklanovairina) [inline index.js:23]: use "console.error"

File: src/controllers/user.controller.js [hunk 28]
```diff
@@ -0,0 +1,45 @@
...
+      res.status(400).json({
+        data: "Registration error",
+      });
+    }
+  }
+
+  async login(req, res) {
+    try {
+      const { name, password } = req.body;
+      const login = await userService.login(name, password);
```
[the remark this answers is not in the document]
Comment (baklanovairina) [reply src/controllers/user.controller.js:37]: Why don't you create a token when logging in?

File: src/service/auth-service.js [hunk 29]
```diff
@@ -0,0 +1,53 @@
...
+
+    const userDto = new UserDto(user);
+    const token = tokenService.generateTokens({ ...userDto });
+    await tokenService.saveToken(userDto.id, token.refreshToken);
+    return { token: token, user: userDto };
+  }
+
+  async login(name, password) {
+    const user = await User.findOne({
+      name,
```
Comment (baklanovairina) [inline src/service/auth-service.js:30]: what for to transfer one parameter to a new line??
